designs/2022-eslint-to-esm/README.md

```diff
@@ -0,0 +1,212 @@
+3. Higher maintenance costs. You need to add a build process, test and publish ESM/CJS code.
+4. The CLIs do not support dual-mode.
+
+In the (very) long run, I would suggest that we eventually go with esm-only, but to give users more time to upgrade, as an intermediate state, we could use dual-mode in ESLint v9:
+
+* eslint v9:
+    - ESLint CLI: esm-only
+    - Node.js API: dual mode
+* eslint 1x (at the point when the majority of the ecosystem is ready to support it, likely to be v10/v11/...):
+    - ESLint CLI: esm-only
+    - Node.js API: esm-only
+
+## Toolings esm supports
+
+* mocha ✅
+* proxyquire ❌
+* nyc ❌
+* eslint-plugin-node ❗️(Not fully supported, but we could use a fork: `eslint-plugin-n`)
+
+## Implementations
+1. ESLint v9
+* package.json + `"type": "module"`
+* package.json exports
+* package.json engines >= Node.js v12
+* remove `"use strict"`
+* convert `"./foo"` to `"./foo/index.js"`
+* convert `"./foo"` to `"./foo.js"`
+* convert `require/module.exports` => `import/export`
+* esm toolings(`nyc` -> `c8`, `proxyquire` -> `esmock`)
+* cjs supports(build `dist/api.cjs` and `dist/use-at-your-own-risk.cjs` using rollup)
+* update docs to use esm
+
+2. ESLint 1x
+* remove package.json exports.cjs (use esm-only)
+* remove build process
+* upgrade blocked deps.
+
+## Documentation
+
+<!--
+    How will this RFC be documented? Does it need a formal announcement
+    on the ESLint blog to explain the motivation?
+-->
+Yes. When ESLint v9 is released, it is recommended to announce that users should upgrade to ESM, CJS may be no longer supported in future major versions.
+
+## Drawbacks
+
+<!--
+    Why should we *not* do this? Consider why adding this into ESLint
+    might not benefit the project or the community. Attempt to think 
+    about any opposing viewpoints that reviewers might bring up. 
+
+    Any change has potential downsides, including increased maintenance
+    burden, incompatibility with other tools, breaking existing user
+    experience, etc. Try to identify as many potential problems with
+    implementing this RFC as possible.
+-->
+1. Compatibility issues, see below.
+
+## Backwards Compatibility Analysis
+
+<!--
+    How does this change affect existing ESLint users? Will any behavior
+    change for them? If so, how are you going to minimize the disruption
+    to existing users?
+-->
+
+1. For users of the ESLint CLI: there is no impact and users can write configuration files using ESM/CJS.
+2. For users of the Node.js API:
+* ESLint v9: Will support ESM/CJS, so there will be no compatibility issues for users.
+* ESLint 1x: CJS will no longer be supported. At this point developers can:
+
+a. convert to esm.
+```diff
+- var { ESLint } = require("eslint");
++ import { ESLint } from "eslint";
+```
+
+b. move to async.
+```diff
+- var { ESLint } = require("eslint");
++ var { ESLint } = await import("eslint");
+```
+
+c. compile ESLint to CJS.
```

Comment:
it's simply not tenable or safe to transpile code one didn't author, so i'd suggest eslint not recommend this at all.